File: autocontext/src/autocontext/runtimes/pi_cli.py
"""Pi CLI runtime — wraps `pi --print` for agent execution.

Uses Pi's non-interactive print mode as a one-shot agent runtime,
capturing output and normalizing into autocontext artifacts.
"""

from __future__ import annotations

import json
import logging
import shutil
import subprocess
import time
from dataclasses import dataclass, field

from autocontext.runtimes.base import AgentOutput, AgentRuntime
from autocontext.runtimes.pi_artifacts import PiExecutionTrace
from autocontext.runtimes.pi_defaults import PI_DEFAULT_TIMEOUT_SECONDS

logger = logging.getLogger(__name__)


@dataclass(slots=True)
class PiCLIConfig:
    """Configuration for the Pi CLI runtime."""

    pi_command: str = "pi"
    model: str = ""
    timeout: float = PI_DEFAULT_TIMEOUT_SECONDS
    json_output: bool = True
    workspace: str = ""
    extra_args: list[str] = field(default_factory=list)


class PiCLIRuntime(AgentRuntime):
    """Agent runtime that invokes the Pi CLI in non-interactive mode.

    Requires the Pi CLI to be installed and accessible on PATH.
    """

    def __init__(self, config: PiCLIConfig | None = None) -> None:
        self._config = config or PiCLIConfig()
        self._pi_path = shutil.which(self._config.pi_command)

    @property
    def available(self) -> bool:
        """Check if the pi CLI is available."""
        return self._pi_path is not None

    def generate(
        self,
        prompt: str,
        system: str | None = None,
        schema: dict | None = None,
    ) -> AgentOutput:
        full_prompt = prompt
        if system:
            full_prompt = f"{system}\n\n{prompt}"
        args = self._build_args(full_prompt)
        return self._invoke(full_prompt, args)

    def revise(
        self,
        prompt: str,
        previous_output: str,
        feedback: str,
        system: str | None = None,
    ) -> AgentOutput:
        revision_prompt = (
            f"Revise the following output based on the judge's feedback.\n\n"
            f"## Original Output\n{previous_output}\n\n"
            f"## Judge Feedback\n{feedback}\n\n"
            f"## Original Task\n{prompt}\n\n"
            "Produce an improved version:"
        )
        full_prompt = revision_prompt
        if system:
            full_prompt = f"{system}\n\n{revision_prompt}"
        args = self._build_args(full_prompt)
        return self._invoke(full_prompt, args)

    def _build_args(self, prompt: str) -> list[str]:
        """Build the pi CLI argument list.

        Uses --print for one-shot mode per Pi's documented interface.
        Workspace is handled via subprocess cwd (Pi has no --workspace flag).
        """
        pi = self._pi_path or self._config.pi_command
        args = [pi, "--print"]

        if self._config.model:
            args.extend(["--model", self._config.model])

        # NOTE: Pi does not have a --workspace CLI flag.
        # Workspace is passed as subprocess cwd instead (see _invoke).

        args.extend(self._config.extra_args)
        args.append(prompt)
        return args

    def _invoke(self, prompt: str, args: list[str]) -> AgentOutput:
        """Execute pi --print and parse the result."""
        logger.info("invoking pi CLI: %s", " ".join(args[:4]) + "...")
        t0 = time.monotonic()

        try:
            result = subprocess.run(
                args,
                capture_output=True,
                text=True,
                timeout=self._config.timeout,
                cwd=self._config.workspace or None,
            )
        except subprocess.TimeoutExpired:
            logger.error("pi CLI timed out after %.0fs", self._config.timeout)
            return AgentOutput(text="", metadata={"error": "timeout"})
        except FileNotFoundError:
            logger.error("pi CLI not found at %r", self._config.pi_command)
            return AgentOutput(text="", metadata={"error": "pi_not_found"})

        duration_ms = int((time.monotonic() - t0) * 1000)

        if result.returncode != 0:
            logger.warning("pi CLI exited with code %d: %s", result.returncode, result.stderr[:200])
            if not result.stdout.strip():
                return AgentOutput(
                    text="",
                    metadata={"error": "nonzero_exit", "exit_code": result.returncode, "stderr": result.stderr[:500]},
                )

        output = self._parse_output(result.stdout, result.returncode)

        # Attach PiExecutionTrace for artifact persistence (AC-224)
        trace = PiExecutionTrace(
            session_id=output.session_id or "",
            prompt_context=prompt,
            raw_output=result.stdout,
            normalized_output=output.text,
            exit_code=result.returncode,
            duration_ms=duration_ms,
            cost_usd=output.cost_usd or 0.0,
            model=output.model or "pi",
        )
        output.metadata["pi_trace"] = trace

        return output

    def _parse_output(self, raw: str, exit_code: int) -> AgentOutput:
        """Parse output from pi --print."""
        if self._config.json_output:
            try:
                data = json.loads(raw)
                text = data.get("result", data.get("output", ""))
                return AgentOutput(
                    text=text,
                    cost_usd=data.get("cost_usd", 0.0),
                    model=data.get("model", "pi"),
                    session_id=data.get("session_id"),
                    metadata={
                        "exit_code": exit_code,
                        "raw_json": data,
                    },
                )
            except (json.JSONDecodeError, TypeError):
                logger.debug("runtimes.pi_cli: suppressed json.JSONDecodeError), TypeError", exc_info=True)

        # Raw text fallback
        return AgentOutput(
            text=raw.strip(),
            cost_usd=0.0,
            model="pi",
            metadata={"exit_code": exit_code},
        )

File: ts/src/runtimes/pi-cli.ts
/**
 * Pi CLI runtime — wraps `pi --print` for agent execution (AC-361).
 * Mirrors Python's autocontext/runtimes/pi_cli.py.
 */

import { execFileSync } from "node:child_process";
import type { AgentOutput } from "./base.js";

export interface PiCLIConfigOpts {
  piCommand?: string;
  model?: string;
  timeout?: number;
  workspace?: string;
  noContextFiles?: boolean;
}

export class PiCLIConfig {
  readonly piCommand: string;
  readonly model: string;
  readonly timeout: number;
  readonly workspace: string;
  readonly noContextFiles: boolean;

  constructor(opts: PiCLIConfigOpts = {}) {
    this.piCommand = opts.piCommand ?? "pi";
    this.model = opts.model ?? "";
    this.timeout = opts.timeout ?? 120.0;
    this.workspace = opts.workspace ?? "";
    this.noContextFiles = opts.noContextFiles ?? false;
  }
}

export class PiCLIRuntime {
  readonly name = "pi-cli";
  private config: PiCLIConfig;

  constructor(config?: PiCLIConfig) {
    this.config = config ?? new PiCLIConfig();
  }

  async generate(opts: { prompt: string; system?: string }): Promise<AgentOutput> {
    const fullPrompt = opts.system ? `${opts.system}\n\n${opts.prompt}` : opts.prompt;
    return this.invoke(fullPrompt);
  }

  async revise(opts: {
    prompt: string;
    previousOutput: string;
    feedback: string;
  }): Promise<AgentOutput> {
    const revisionPrompt =
      `Revise the following output based on the judge's feedback.\n\n` +
      `## Original Output\n${opts.previousOutput}\n\n` +
      `## Judge Feedback\n${opts.feedback}\n\n` +
      `## Original Task\n${opts.prompt}\n\n` +
      `Produce an improved version:`;
    return this.invoke(revisionPrompt);
  }

  parseOutput(raw: string): AgentOutput {
    const trimmed = raw.trim();
    if (!trimmed) return { text: "", metadata: {} };
    return { text: trimmed, metadata: {} };
  }

  private invoke(prompt: string): AgentOutput {
    const args = ["--print"];
    if (this.config.model) {
      args.push("--model", this.config.model);
    }
    if (this.config.noContextFiles) {
      args.push("--no-context-files");
    }

    try {
      const stdout = execFileSync(this.config.piCommand, args, {
        input: prompt,
        timeout: this.config.timeout * 1000,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
        cwd: this.config.workspace || undefined,
      });
      return this.parseOutput(stdout);
    } catch (err: unknown) {
      const error = err as { code?: string; message?: string };
      if (error.code === "ETIMEDOUT") {
        return { text: "", metadata: { error: "timeout" } };
      }
      return { text: "", metadata: { error: error.message ?? "unknown" } };
    }
  }
}
